mtags/src/main/scala-3/scala/meta/internal/pc/MetalsDriver.scala

```diff
@@ -0,0 +1,42 @@
+package scala.meta.internal.pc
+
+import java.net.URI
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta.internal.mtags.MD5
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.interactive.SourceTree
+import dotty.tools.dotc.reporting.Diagnostic
+import dotty.tools.dotc.util.SourceFile
+
+class MetalsDriver(
+    override val settings: List[String]
+) extends InteractiveDriver(settings):
+
+  private val lastCompiled: TrieMap[URI, String] = TrieMap.empty
+
+  private def alreadyCompiled(uri: URI, hash: String): Boolean =
+    lastCompiled
+      .get(uri)
+      .map(cache => cache == hash)
+      .getOrElse(false)
+
+  override def run(uri: URI, source: SourceFile): List[Diagnostic] =
+    val content = source.content.mkString
+    val contentHash = MD5.compute(content)
+    if alreadyCompiled(uri, contentHash) then Nil
+    else
+      lastCompiled.put(uri, contentHash)
+      super.run(uri, source)
```

Comment:
Shouldn't we also override `def currectCtx`?

Reply:
I think we can use the `parent`'s `currentCtx` implementation (so I think we don't need to override `currentCtx`), what do you think about it?

Reply:
I think it will work incorrectly in  the following case:
- call pc for `A.scala` (compilation happens) - `driver.currentCtx` refers to the ctx from `A.scala`
- call pc for `B.scala` (compilation happens) - `driver.currentCtx` refers to the ctx from `B.scala`
- switch back on `A.scala` without changes, call pc - `driver.currentCtx` refers to the ctx from `B.scala`

Reply:
Hm, yeah, that should be a problem :+1: (maybe `run` should update the `currentCtx` anyway even MetalsDriver doesn't run compile).

I tried to produce some problems, but couldn't. Did you find any problem with the lack of `currentCtx` implementation?
I realized `driver.currentCtx.source` always returns `<no file>` and `driver.currentCtx.source.tpdTree` returns `EmptyTree`. Also, I skimmed through the `InteractiveDriver` and `Run`, but it looks like no one assigns the given source to the `currentCtx`. Maybe `currentCtx` doesn't refer to any specific files?

Reply:
It's being used in `InferredTypeProvider` maybe you can check inserting type if it works properly?

Reply:
@tanishiking I think it would be better to override it. I've disabled `auto save`  in vscode and got the situation at autoimports here: https://github.com/scalameta/metals/blob/15b8787c0f08520ef3bf54071b6ed8d318d9a703/mtags/src/main/scala-3/scala/meta/internal/pc/AutoImportsProvider.scala#L37-L43

where the request uri wasn't equal to `driver.currentCtx.run.units.head`

Reply:
@dos65 Great! I could reproduce it :+1:

Reply:
Tried to override `currentCtx` so that it returns the context connected to the last `run`ed compilation unit
https://github.com/scalameta/metals/pull/4225/commits/553e2b33ecc473849466db0afc4b51ee9162d9f5

This implementation solved the wrong context issue, but something funky started happening, for example when I run code completion. It looks like, under the cached context (whose `runId = 3`) the symbol will be invalid in the new run (where `runId = 4`).

I'm considering, maybe we should override `currentCtx` like

```scala
override def currentCtx =
  val ctx0 = super.currentCtx
  ctx0.setCompilationUnit(<last compiled compilation unit for the file>)
```


```
SEVERE: assertion failed: denotation module class C$ invalid in run 3. ValidFor: Period(1..5, run = 4)
java.lang.AssertionError: assertion failed: denotation module class C$ invalid in run 3. ValidFor: Period(1..5, run = 4)
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.updateValidity(Denotations.scala:722)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.bringForward(Denotations.scala:747)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.toNewRun$1(Denotations.scala:806)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:880)
        at dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:120)
        at dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:114)
        at dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:107)
        at dotty.tools.dotc.core.Symbols$.toDenot(Symbols.scala:497)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.updateValidity(Denotations.scala:721)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.bringForward(Denotations.scala:747)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.toNewRun$1(Denotations.scala:806)
        at dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:880)
        at dotty.tools.dotc.core.Types$NamedType.computeDenot(Types.scala:2256)
        at dotty.tools.dotc.core.Types$NamedType.denot(Types.scala:2216)
        at dotty.tools.dotc.ast.Trees$DenotingTree.denot(Trees.scala:256)
        at dotty.tools.dotc.ast.Trees$Tree.symbol(Trees.scala:145)
        at scala.meta.internal.pc.MetalsInteractive$.contextOfPath(MetalsInteractive.scala:56)
        at scala.meta.internal.pc.MetalsInteractive$.contextOfPath(MetalsInteractive.scala:34)
        at scala.meta.internal.pc.MetalsInteractive$.contextOfPath(MetalsInteractive.scala:34)
        at scala.meta.internal.pc.completions.CompletionsProvider.completions(CompletionsProvider.scala:59)
```

Reply:
> I'm considering, maybe we should override currentCtx

Realized it doesn't work, we have to set compilation units to the `ctx.run` but it's not accessible outside of `Run`.
Hmmm, maybe we should do this in compiler as Scala2 does 🤔

Reply:
> if there will be a sequential hover calls for the same source without changes

Ohhh, that's a good idea :+1:

Reply:
@dos65 I implemented like
> if there will be a sequential hover calls for the same source without changes
and looks like the problem has gone :tada: